Indexing/CodeTree.hpp

```diff
@@ -188,65 +191,76 @@ class CodeTree
-    inline bool isFail() const { return !data(); }
-    inline bool isLitEnd() const { return instrPrefix()==LIT_END; }
-    inline bool isSearchStruct() const { return instrPrefix()==SUFFIX_INSTR && instrSuffix()==SEARCH_STRUCT; }
-    inline bool isCheckFun() const { return instrPrefix()==SUFFIX_INSTR && instrSuffix()==CHECK_FUN; }
-    inline bool isCheckGroundTerm() const { return instrPrefix()==CHECK_GROUND_TERM; }
+    inline bool isSuccess() const { return _instruction()==SUCCESS_OR_FAIL && _data<void>(); }
+    inline bool isFail() const { return !_data<void>(); }
+    inline bool isLitEnd() const { return _instruction()==LIT_END; }
+    inline bool isSearchStruct() const { return _instruction()==SEARCH_STRUCT; }
+    inline bool isCheckFun() const { return _instruction()==CHECK_FUN; }
+    inline bool isCheckGroundTerm() const { return _instruction()==CHECK_GROUND_TERM; }
 
     inline Term* getTargetTerm() const
     {
       ASS(isCheckGroundTerm());
-      return reinterpret_cast<Term*>(data()&~static_cast<size_t>(CHECK_GROUND_TERM));
+      return _data<Term>();
     }
 
-    inline void* getSuccessResult() { ASS(isSuccess()); return _result; }
+    template<class T> inline T* getSuccessResult() { ASS(isSuccess()); return _data<T>(); }
 
     inline ILStruct* getILS()
     {
       ASS(isLitEnd());
-      return reinterpret_cast<ILStruct*>(data()&~static_cast<size_t>(LIT_END));
+      return _data<ILStruct>();
     }
     inline const ILStruct* getILS() const
     {
-      return const_cast<CodeOp*>(this)->getILS();
+      return _data<ILStruct>();
     }
 
     SearchStruct* getSearchStruct() const;
 
-    inline InstructionPrefix instrPrefix() const { return static_cast<InstructionPrefix>(_info.prefix); }
-    inline InstructionSuffix instrSuffix() const
-    {
-      ASS_EQ(instrPrefix(), SUFFIX_INSTR);
-      return static_cast<InstructionSuffix>(_info.suffix);
-    }
-
-    inline unsigned arg() const { return _info.arg; }
     inline CodeOp* alternative() const { return _alternative; }
     inline CodeOp*& alternative() { return _alternative; }
 
     inline void setAlternative(CodeOp* op) { ASS_NEQ(op, this); _alternative=op; }
-    inline void setLongInstr(InstructionSuffix i) { _info.prefix=SUFFIX_INSTR; _info.suffix=i; }
 
-    void makeFail() { _data=0; }
+    void makeFail() { _setData<void>(0); }
 
     friend std::ostream& operator<<(std::ostream& out, const CodeOp& op);
 
+    static constexpr unsigned
+      INSTRUCTION_BITS_START = 0,
+      INSTRUCTION_BITS_END = INSTRUCTION_BITS_START + 3,
+      ARG_BITS_START = INSTRUCTION_BITS_END,
+      ARG_BITS_END = CHAR_BIT * sizeof(uint64_t),
+      DATA_BITS_START = 0,
+      DATA_BITS_END = CHAR_BIT * sizeof(void *);
+
+    static_assert(sizeof(void *) <= sizeof(uint64_t), "must be able to fit a pointer into a 64-bit integer");
+    static_assert(SEARCH_STRUCT < 8, "must be able to squash instructions into 3 bits");
+    static_assert(alignof(Term) == 8);
```

Comment:
Strictly speaking `alignof(Term)` should be a *multiple* of 8. But this is probably fine, and we'll get a nice compile-time failure if it isn't.